### PR TITLE
[README] Update installation instructions for linear solvers

### LINKED ISSUE #247: Linking Ipopt with linear solvers
There are a lot of issues relating to linear solvers with Ipopt. #236 was a start at providing some documentation on how to use Ipopt with different solvers: https://github.com/jump-dev/Ipopt.jl#custom-installation. However, most are untested and I don't have a Windows machine. I've opened this issue to track which solver-platform combinations we have instructions for.

If you have clarifications for existing instructions, or if you have instructions for a missing solver-platform pair, please open a pull request.

Here are solver/platform combinations I have tested and got working so far:

* Pardiso Project
  - [ ] Windows
  - [x] Mac
  - [x] Linux  (Related issues: #106, #177)
* MKL Pardiso
  - [ ] Windows
  - [ ] Mac
  - [ ] Linux (Related issues: #195)
* HSL for Ipopt (MA27)
  - [ ] Windows (Related issues: #229)
  - [x] Mac
  - [x] Linux
* HSL (MA86, MA97)
  - [ ] Windows (Related issues: #192)
  - [x] Mac (Related issues: #197, #213)
  - [x] Linux

### LINKED ISSUE #360: pointer being freed was not allocated
Hi! I am using an Apple MacBook Air (M1, 2020), and in my `~/.zshrc`, I have a line like this
```
export DYLD_LIBRARY_PATH=/Users/henrique/coinbrew/Ipopt/dist/lib
```
for exporting the path to the Ipopt that I built with [coinbrew](https://github.com/coin-or/coinbrew), and which allows me to use different linear solvers besides MA27.

I have been using it without issues, with Julia itself having been installed with `brew install julia` (https://github.com/Homebrew/homebrew-core/blob/master/Formula/julia.rb).

```julia
julia> versioninfo()
Julia Version 1.8.5
Commit 17cfb8e65e* (2023-01-08 06:45 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin22.1.0)
  CPU: 8 × Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 1 on 4 virtual cores
Environment:
  DYLD_LIBRARY_PATH = /Users/henrique/knitro/lib:/Users/henrique/coinbrew/Ipopt/dist/lib

julia> using Ipopt

julia>
```

As you can see in the snippet above, there are no issues loading Ipopt.

However, if I install Julia from the official `.dmg` file or with [juliaup](https://github.com/JuliaLang/juliaup), I get this error:
```julia
julia> versioninfo()
Julia Version 1.8.5
Commit 17cfb8e65ea (2023-01-08 06:45 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.5.0)
  CPU: 8 × Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 1 on 4 virtual cores
Environment:
  DYLD_LIBRARY_PATH = /Users/henrique/knitro/lib:/Users/henrique/coinbrew/Ipopt/dist/lib

julia> using Ipopt
[ Info: Precompiling Ipopt [b6b21f68-93f8-5de0-b562-5493be1d77c9]
julia(61172,0x1e6005b40) malloc: *** error for object 0x10fff07d0: pointer being freed was not allocated
julia(61172,0x1e6005b40) malloc: *** set a breakpoint in malloc_error_break to debug

signal (6): Abort trap: 6
in expression starting at /Users/henrique/.julia/packages/Ipopt/Re4kE/src/Ipopt.jl:8
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
Allocations: 2906 (Pool: 2895; Big: 11); GC: 0
ERROR: Failed to precompile Ipopt [b6b21f68-93f8-5de0-b562-5493be1d77c9] to /Users/henrique/.julia/compiled/v1.8/Ipopt/jl_0UwpOD.
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] compilecache(pkg::Base.PkgId, path::String, internal_stderr::IO, internal_stdout::IO, keep_loaded_modules::Bool)
   @ Base ./loading.jl:1707
 [3] compilecache
   @ ./loading.jl:1651 [inlined]
 [4] _require(pkg::Base.PkgId)
   @ Base ./loading.jl:1337
 [5] _require_prelocked(uuidkey::Base.PkgId)
   @ Base ./loading.jl:1200
 [6] macro expansion
   @ ./loading.jl:1180 [inlined]
 [7] macro expansion
   @ ./lock.jl:223 [inlined]
 [8] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:1144
```

I don't know if this is an issue with Ipopt or with Julia, but I decided to post it here first. I had a look at the [Homebrew Formula for Julia](https://github.com/Homebrew/homebrew-core/blob/master/Formula/julia.rb) to try and understand if they are doing something special there for not running into this error, but I am quite out of my depth...

**Edit:** Oh, and another thing worth noting is that, if I comment out that export line from my `~/.zshrc`, this is not an issue.

### PR DESCRIPTION
**Pull request by odow**
@amontoison has done an incredible amount of work to try and simplify the installation process for HSL in Julia. The result is https://licences.stfc.ac.uk/product/julia-hsl, an officially supported pre-built binary distribution of _all_ the HSL solvers.

@amontoison are these instructions correct? I don't have an academic account, so I can't actually test this 😆 

Closes #360 
Closes #247

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -228,147 +228,42 @@ model = Model(Ipopt.Optimizer)
...
+set_attribute(model, "hsllib", HSL_jll.libhsl_path)
+set_attribute(model, "linear_solver", "ma86")
+```
+
+### Pardiso
+
+Download Pardiso from [https://www.pardiso-project.org](https://www.pardiso-project.org).
+Save the shared library somewhere, and record the filename.
+
+Then, use Pardiso by setting the `parsidolib` and `linear_solver` attributes:
```
Comment (odow) [inline README.md:257]:
```suggestion
Then, use Pardiso by setting the `pardisolib` and `linear_solver` attributes:
```

File: README.md [hunk 2]
```diff
@@ -228,147 +228,42 @@ model = Model(Ipopt.Optimizer)
...
+```julia
+using JuMP, Ipopt
+model = Model(Ipopt.Optimizer)
+set_attribute(model, "pardisolib", "/full/path/to/libparsido")
```
Comment (amontoison) [inline README.md:261]:
```suggestion
set_attribute(model, "pardisolib", "/full/path/to/libpardiso")
```

File: README.md [hunk 3]
```diff
@@ -228,147 +228,42 @@ model = Model(Ipopt.Optimizer)
...
-Alternatively you can use [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/install)
-and follow the Linux instructions.
-
-### Pardiso (MKL)
-
-Currently untested on all platforms. If you have instructions that work, please
-open an issue.
+### Windows
+
+Changing the linear solver dynamically may not work on Windows. Instead, you can
```
Comment (amontoison) [inline README.md:267]: For information, HSL_jll.jl is precompiled on Windows platforms. Is it an issue with the source code of Ipopt?

**Comment by amontoison** [comment]
@AndrewLister-STFC
As the main maintener of Ipopt.jl, is it possible to authorize Oscar to download JuliaHSL?

**Comment by amontoison** [comment]
@odow
The instructions are correct 🙂

File: README.md (hunk 3 above)
Comment (odow) [reply README.md:267]: Hmm. I've had a few reports. Should we just delete this section and see if people complain?

Comment (odow) [reply README.md:267]: https://github.com/jump-dev/Ipopt.jl/issues/247#issuecomment-731910914

Comment (amontoison) [reply README.md:267]: Hum, I was not aware of this issue. Let's keep this section and wait the feedbacks of the users.

If it's not working, we could link Ipopt with a dummy JuliaHSL 2.0. We just need to wait dummy routines for the symbols of HSL solvers used in Ipopt.

**Review by amontoison — APPROVED** [review]

File: README.md (hunk 3 above)
Comment (amontoison) [reply README.md:267]: @odow 
I tested on Windows and we don't have any issue if we use `HSL_jll.jl` with `Ipopt.jl`.
```julia
using HSL_jll, Ipopt, JuMP


println(HSL_jll.libhsl_path)

# An optimization problem
model = Model(Ipopt.Optimizer)
@variable(model, x)
@objective(model, Min, (x - 2)^2)

# Load the HSL solvers
set_attribute(model, "hsllib", HSL_jll.libhsl_path)

# Use the linear solver MA27
set_attribute(model, "linear_solver", "ma27")
optimize!(model)

# Use the linear solver MA57
set_attribute(model, "linear_solver", "ma57")
optimize!(model)

# Use the linear solver MA77
set_attribute(model, "linear_solver", "ma77")
optimize!(model)

# Use the linear solver MA86
set_attribute(model, "linear_solver", "ma86")
optimize!(model)

# Use the linear solver MA97
set_attribute(model, "linear_solver", "ma97")
optimize!(model)
```

```julia
     Testing Running tests...
D:\a\HSL_jll.jl\HSL_jll.jl\override\bin\x86_64-w64-mingw32-libgfortran5\libhsl.dll

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt version 3.14.10, running with linear solver ma27.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        1

Total number of variables............................:        1
                     variables with only lower bounds:        0
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:        0
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  4.0000000e+00 0.00e+00 4.00e+00  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  0.0000000e+00 0.00e+00 0.00e+00  -1.0 2.00e+00    -  1.00e+00 1.00e+00f  1

Number of Iterations....: 1

                                   (scaled)                 (unscaled)
Objective...............:   0.0000000000000000e+00    0.0000000000000000e+00
Dual infeasibility......:   0.0000000000000000e+00    0.0000000000000000e+00
Constraint violation....:   0.0000000000000000e+00    0.0000000000000000e+00
Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   0.0000000000000000e+00    0.0000000000000000e+00


Number of objective function evaluations             = 2
Number of objective gradient evaluations             = 2
Number of equality constraint evaluations            = 0
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 0
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 1
Total seconds in IPOPT                               = 0.156

EXIT: Optimal Solution Found.
This is Ipopt version 3.14.10, running with linear solver ma57.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        1

Total number of variables............................:        1
                     variables with only lower bounds:        0
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:        0
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  4.0000000e+00 0.00e+00 4.00e+00  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  0.0000000e+00 0.00e+00 0.00e+00  -1.0 2.00e+00    -  1.00e+00 1.00e+00f  1

Number of Iterations....: 1

                                   (scaled)                 (unscaled)
Objective...............:   0.0000000000000000e+00    0.0000000000000000e+00
Dual infeasibility......:   0.0000000000000000e+00    0.0000000000000000e+00
Constraint violation....:   0.0000000000000000e+00    0.0000000000000000e+00
Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   0.0000000000000000e+00    0.0000000000000000e+00


Number of objective function evaluations             = 2
Number of objective gradient evaluations             = 2
Number of equality constraint evaluations            = 0
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 0
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 1
Total seconds in IPOPT                               = 0.000

EXIT: Optimal Solution Found.
This is Ipopt version 3.14.10, running with linear solver ma77.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        1

Total number of variables............................:        1
                     variables with only lower bounds:        0
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:        0
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  4.0000000e+00 0.00e+00 4.00e+00  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  0.0000000e+00 0.00e+00 0.00e+00  -1.0 2.00e+00    -  1.00e+00 1.00e+00f  1

Number of Iterations....: 1

                                   (scaled)                 (unscaled)
Objective...............:   0.0000000000000000e+00    0.0000000000000000e+00
Dual infeasibility......:   0.0000000000000000e+00    0.0000000000000000e+00
Constraint violation....:   0.0000000000000000e+00    0.0000000000000000e+00
Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   0.0000000000000000e+00    0.0000000000000000e+00


Number of objective function evaluations             = 2
Number of objective gradient evaluations             = 2
Number of equality constraint evaluations            = 0
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 0
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 1
Total seconds in IPOPT                               = 0.000

EXIT: Optimal Solution Found.
This is Ipopt version 3.14.10, running with linear solver ma86.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        1

Total number of variables............................:        1
                     variables with only lower bounds:        0
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:        0
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  4.0000000e+00 0.00e+00 4.00e+00  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  0.0000000e+00 0.00e+00 0.00e+00  -1.0 2.00e+00    -  1.00e+00 1.00e+00f  1

Number of Iterations....: 1

                                   (scaled)                 (unscaled)
Objective...............:   0.0000000000000000e+00    0.0000000000000000e+00
Dual infeasibility......:   0.0000000000000000e+00    0.0000000000000000e+00
Constraint violation....:   0.0000000000000000e+00    0.0000000000000000e+00
Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   0.0000000000000000e+00    0.0000000000000000e+00


Number of objective function evaluations             = 2
Number of objective gradient evaluations             = 2
Number of equality constraint evaluations            = 0
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 0
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 1
Total seconds in IPOPT                               = 0.000

EXIT: Optimal Solution Found.
This is Ipopt version 3.14.10, running with linear solver ma97.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        1

Total number of variables............................:        1
                     variables with only lower bounds:        0
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:        0
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  4.0000000e+00 0.00e+00 4.00e+00  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  0.0000000e+00 0.00e+00 0.00e+00  -1.0 2.00e+00    -  1.00e+00 1.00e+00f  1

Number of Iterations....: 1

                                   (scaled)                 (unscaled)
Objective...............:   0.0000000000000000e+00    0.0000000000000000e+00
Dual infeasibility......:   0.0000000000000000e+00    0.0000000000000000e+00
Constraint violation....:   0.0000000000000000e+00    0.0000000000000000e+00
Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
Overall NLP error.......:   0.0000000000000000e+00    0.0000000000000000e+00


Number of objective function evaluations             = 2
Number of objective gradient evaluations             = 2
Number of equality constraint evaluations            = 0
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 0
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 1
Total seconds in IPOPT                               = 0.000

EXIT: Optimal Solution Found.
```

Comment (amontoison) [reply README.md]: I ran the same code with the two versions of `HSL_jll.jl` (OpenBLAS32 -- LBT) on the following platforms : 
- Ubuntu
- macOS
- Windows
- FreeBSD
- Linux ARMv8
- MacOS M1

Everything is working like a charm :muscle:

Comment (AndrewLister-STFC) [reply README.md]: @antoine-guillou Do you already have JuliaHSL? https://licences.stfc.ac.uk/product/julia-hsl
I think `HSL_jll.is_available()` should tell you if it's finding the right `libhsl.dll`.

If you're not an academic (so need this to work with "CoinHSL Archive"), I'm not sure how to get HSL_jll to use the right library. You may be able to just swap the dll in the path printed with the one you got from the STFC IP Store?

**Pull request merged**